src/java/org/apache/cassandra/index/sai/QueryContext.java

```diff
@@ -181,18 +181,16 @@ public Bits bitsetForShadowedPrimaryKeys(SegmentMetadata metadata, PrimaryKeyMap
         if (ignoredOrdinals == null)
             return null;
 
-        return new IgnoringBits(graph, ignoredOrdinals, metadata);
+        return new IgnoringBits(ignoredOrdinals, metadata);
     }
 
     private static class IgnoringBits implements Bits
     {
-        private final CassandraOnDiskHnsw graph;
         private final Set<Integer> ignoredOrdinals;
         private final int length;
 
-        public IgnoringBits(CassandraOnDiskHnsw graph, Set<Integer> ignoredOrdinals, SegmentMetadata metadata)
+        public IgnoringBits(Set<Integer> ignoredOrdinals, SegmentMetadata metadata)
         {
-            this.graph = graph;
             this.ignoredOrdinals = ignoredOrdinals;
             this.length = 1 + metadata.toSegmentRowId(metadata.maxSSTableRowId);
```

Comment:
If I understand the purpose of this implementation of `Bits`, it's supposed to take in a set of graph ordinals, which are precisely the ordinals to be ignored at search time. Every other ordinal, the corresponding bit is set.

If my understanding is correct, then my question: why have we been specifying `length` to be this value rather than `graph.size()`?
- ordinals are not necessarily in 1-to-1 correspondence with segment row id's
- even if they were, [this discussion](https://github.com/datastax/cassandra/pull/722#discussion_r1316698880) would indicate that the max segment row id can exceed the last ordinal in the graph (since the first indexed row in the segment may not necessarily have segment row id  = 0)

Reply:
You're right.  The search path never calls length() and I don't think anything else does either, but if we're going to implement it we should do it correctly.  I'll open a separate PR for this.